### DISCUSSION TITLE
ERROR    Incompatible Python version: 3.11.3 required 3.[9, 10]

### DISCUSSION THREAD

**Discussion by Rizzlord** [vladmandic/sdnext#1685] [General]
I use garuda linux its an Arch Linux, and it comes with python 3.11 out of the box, every command to download or update python will end up installing 3.11... So i went to install 3.10 vom the AUR, but it still dont work. Any help? im getting crazy here...

**Comment by brknsoul** [comment]
From ArchWiki
>  Python packages are installed in a versioned site-packages directory (/usr/lib/pythonX.Y/site-packages if system-wide, or ~/.local/lib/pythonX.Y/site-packages/ if per-user, where X.Y is a version like "3.11").

run `which -a python` to see where python is installed. If you're unsure of the version, you could simply run found instances of python with `/path/to/python --version`

- Delete the `venv` folder.
- run `export PYTHON=/path/to/python3.10`
- run `webui.sh`

**Comment by Rizzlord** [comment]
> From ArchWiki
> 
> > Python packages are installed in a versioned site-packages directory (/usr/lib/pythonX.Y/site-packages if system-wide, or ~/.local/lib/pythonX.Y/site-packages/ if per-user, where X.Y is a version like "3.11").
> 
> run `which -a python` to see where python is installed. If you're unsure of the version, you could simply run found instances of python with `/path/to/python --version`
> 
> * Delete the `venv` folder.
> * run `export PYTHON=/path/to/python3.10`
> * run `webui.sh`

did not work even tho i checked if its changed it will still use 3.11

**Reply by brknsoul** [reply]
Sorry, that's the extent of my assistance. :-P
I'm a Windows addict. I've only casually delved into linux, and even then, Ubuntu.

Perhaps you could alias `python`?

**Comment by artemesart** [comment]
> I use garuda linux its an Arch Linux, and it comes with python 3.11 out of the box, every command to download or update python will end up installing 3.11... So i went to install 3.10 vom the AUR, but it still dont work. Any help? im getting crazy here...

Hello! I had the same problem and I solved it this way: 
Installed Python 3.10 from AUR
`yay -S python310`

then in the webui.sh file in the block
` #python3 executable `
replaced the line 
`python_cmd="python3" ` to 
`python_cmd="python3.10"`

After that, I ran 
`$ bash webui.sh`

The first time I ran it I got the error "cannot import name 'deprecated' from 'typing_extensions'"
I just restarted and it worked fine 
`$ bash webui.sh`

**Reply by Rizzlord** [reply]
Create and activate python venv
Setting OneAPI environment

:: initializing oneAPI environment ...
webui.sh: BASH_VERSION = 5.1.16(1)-release
args: Using "$@" for setvars.sh arguments: --use-rocm
:: compiler -- latest
:: mkl -- latest
:: tbb -- latest
:: oneAPI environment initialized ::

Launching launch.py...
Traceback (most recent call last):
File "/run/media/rizzlord/Gigantus/automatic/launch.py", line 8, in <module>
import installer
File "/run/media/rizzlord/Gigantus/automatic/installer.py", line 12, in <module>
import pkg_resources
ModuleNotFoundError: No module named 'pkg_resources'

:D

**Reply by artemesart** [reply]
try to remove the venv folder and try again.

**Reply by Rizzlord** [reply]
its installing... last question would be, how and where do i put arguments, and how should i start it with them? i used automatic before on windows, there was a web-user.bat, now im on linux and how would i do that here? BTW u are awesome, i worked 2 days on it...

EDIT: ImportError: cannot import name 'deprecated' from 'typing_extensions'
(/run/media/rizzlord/Gigantus/automatic/venv/lib/python3.10/site-packages/typing_extensions.py)

EDIT2: works now. Still need to know how to add arguments.

EDIT3: everything seems fine, but when i hit generate, nothing happens.

01:18:09-510471 INFO     ControlNet v1.1.232
ControlNet v1.1.232
01:18:10-272858 INFO     Loading UI theme: name=black-orange style=Auto
Running on local URL:  http://127.0.0.1:7860
01:18:11-151682 INFO     Local URL: http://127.0.0.1:7860/
01:18:11-152581 INFO     Initializing middleware
01:18:11-241629 INFO     [AgentScheduler] Task queue is empty
01:18:11-242389 INFO     [AgentScheduler] Registering APIs
Loading weights: /run/media/rizzlord/Gigantus/automatic/models/Stable-diffusion/v1-5-pruned-emaonly.safetens…
01:18:11-842975 INFO     Setting Torch parameters: dtype=torch.float16 vae=torch.float16 unet=torch.float16
LatentDiffusion: Running in eps-prediction mode
DiffusionWrapper has 859.52 M params.

**Reply by brknsoul** [reply]
You're trying to generate too early. 

Wait until you see something like this in the console;
```
Startup time: 12.3s (scripts=3.9s onchange=0.3s ui-txt2img=0.3s ui-img2img=0.1s ui-settings=0.1s ui-extensions=5.7s
ui-defaults=0.1s launch=0.3s app-started=0.7s checkpoint=0.8s)
```

If you use the `--autolaunch` option, then the browser will auto-open the page when loading is complete.

As to how to add command line options;
`./webui.sh --autolaunch`

Note that all of the old optimisation command line options are now under Settings > Optimization.

**Reply by artemesart** [reply]
some parameters can be set on the command line, e.g. 
`$ bash webui.sh --medvram`

Most of the settings (e.g. xformers) have been moved to the Settings tab 

>EDIT3: everything seems fine, but when i hit generate, nothing happens.

really looks like everything's fine. I don't know what the problem is.
you can try running with the flag 
`$ bash webui.sh --debug`
There should be more information there.

**Reply by Rizzlord** [reply]
reinstalled everything
02:14:31-092507 INFO     Server arguments: []
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /run/media/rizzlord/Gigantus/automatic/launch.py:190 in <module>                                 │
│                                                                                                  │
│   187 │   # installer.log.debug(f"Args: {vars(args)}")                                           │
│   188 │   logging.disable(logging.NOTSET if args.debug else logging.DEBUG)                       │
│   189 │                                                                                          │
│ ❱ 190 │   instance = start_server(immediate=True, server=None)                                   │
│   191 │   while True:                                                                            │
│   192 │   │   try:                                                                               │
│   193 │   │   │   alive = instance.thread.is_alive()                                             │
│                                                                                                  │
│ /run/media/rizzlord/Gigantus/automatic/launch.py:136 in start_server                             │
│                                                                                                  │
│   133 │   server = importlib.util.module_from_spec(module_spec)                                  │
│   134 │   installer.log.debug(f'Starting module: {server}')                                      │
│   135 │   installer.log.info(f"Server arguments: {sys.argv[1:]}")                                │
│ ❱ 136 │   module_spec.loader.exec_module(server)                                                 │
│   137 │   if args.test:                                                                          │
│   138 │   │   installer.log.info("Test only")                                                    │
│   139 │   │   server.wants_restart = False                                                       │
│ in exec_module:883                                                                               │
│ in _call_with_frames_removed:241                                                                 │
│                                                                                                  │
│ /run/media/rizzlord/Gigantus/automatic/webui.py:35 in <module>                                   │
│                                                                                                  │
│    32 startup_timer.record("torch")                                                              │
│    33                                                                                            │
│    34 errors.log.debug('Loading Gradio')                                                         │
│ ❱  35 from fastapi import FastAPI # pylint: disable=W0611,C0411                                  │
│    36 import gradio # pylint: disable=W0611,C0411                                                │
│    37 startup_timer.record("gradio")                                                             │
│    38 errors.install([gradio])                                                                   │
│                                                                                                  │
│ /run/media/rizzlord/Gigantus/automatic/venv/lib/python3.10/site-packages/fastapi/__init__.py:7   │
│ in <module>                                                                                      │
│                                                                                                  │
│    4                                                                                             │
│    5 from starlette import status as status                                                      │
│    6                                                                                             │
│ ❱  7 from .applications import FastAPI as FastAPI                                                │
│    8 from .background import BackgroundTasks as BackgroundTasks                                  │
│    9 from .datastructures import UploadFile as UploadFile                                        │
│   10 from .exceptions import HTTPException as HTTPException                                      │
│                                                                                                  │
│ /run/media/rizzlord/Gigantus/automatic/venv/lib/python3.10/site-packages/fastapi/applications.py │
│ :16 in <module>                                                                                  │
│                                                                                                  │
│    13 │   Union,                                                                                 │
│    14 )                                                                                          │
│    15                                                                                            │
│ ❱  16 from fastapi import routing                                                                │
│    17 from fastapi.datastructures import Default, DefaultPlaceholder                             │
│    18 from fastapi.exception_handlers import (                                                   │
│    19 │   http_exception_handler,                                                                │
│                                                                                                  │
│ /run/media/rizzlord/Gigantus/automatic/venv/lib/python3.10/site-packages/fastapi/routing.py:22   │
│ in <module>                                                                                      │
│                                                                                                  │
│     19 │   Union,                                                                                │
│     20 )                                                                                         │
│     21                                                                                           │
│ ❱   22 from fastapi import params                                                                │
│     23 from fastapi._compat import (                                                             │
│     24 │   ModelField,                                                                           │
│     25 │   Undefined,                                                                            │
│                                                                                                  │
│ /run/media/rizzlord/Gigantus/automatic/venv/lib/python3.10/site-packages/fastapi/params.py:6 in  │
│ <module>                                                                                         │
│                                                                                                  │
│     3 from typing import Any, Callable, Dict, List, Optional, Sequence, Union                    │
│     4                                                                                            │
│     5 from pydantic.fields import FieldInfo                                                      │
│ ❱   6 from typing_extensions import Annotated, deprecated                                        │
│     7                                                                                            │
│     8 from ._compat import PYDANTIC_V2, Undefined                                                │
│     9                                                                                            │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
ImportError: cannot import name 'deprecated' from 'typing_extensions'
(/run/media/rizzlord/Gigantus/automatic/venv/lib/python3.10/site-packages/typing_extensions.py)

**Reply by brknsoul** [reply]
The fix for that is to simply run webui again.

**Comment by Rizzlord** [comment]
Fixed, had to install an older torch version "You are running torch 1.13.1+rocm5.2.
The program is tested to work with torch 2.0.0.
To reinstall the desired version, run with commandline flag --reinstall-torch.
Beware that this will cause a lot of large files to be downloaded, as well as
there are reports of issues with training tab on the latest version." but, still works, dont wanna try to upgrade, maybe itl break again.

for all having problems on arch linux here is the command i used: "pip install torch==1.13.1+rocm5.2 torchvision==0.14.1+rocm5.2 --extra-index-url https://download.pytorch.org/whl/rocm5.2 --upgrade --force-reinstall"

**Comment by jgbarah** [comment]
I'm having the same problem with my Debian, which has Python 3.11.2.

Actually, it seems the installer is already prepared to work with 3.11. All I had to do is to use the `--experimental` flag (and this should work for any Python 3.11.x):

```
./webui.sh --experimental
```

The only problem seems to be that it cannot load the rembg module.

**Reply by jgbarah** [reply]
Well, that was not right. Even when it appeared to be working, when I tried to produce an image, it failed with a runtime error, apparently something missing in torch. Maybe an issue with the torch version?

So, instead of using `--experimental` I changed the `installer.py` file, adding 11 to the list of supported Python minor versions:

``` 
supported_minors = [9, 10, 11]
```

Now, it seems to work, at least producing images from a prompt (launching without the `--experimental` flag). The problem with the rembg module persists.

**Reply by xedis** [reply]
remove any existing venv
```shell
rm -rf ./venv
```
manually create the python3.10 venv
```shell
python3.10 -m venv venv
```

Start webui
```shell
./webui.sh --debug
``` 
```shell
###profit
```
